# Data Cleaning

Data Cleaning for the Dataset for [Video Game Sales and Ratings](https://www.kaggle.com/datasets/kendallgillies/video-game-sales-and-ratings)

## Import and Read the Dataset

In [1]:
import numpy as np
import pandas as pd

raw_df = pd.read_csv("Video_Game_Sales_as_of_Jan_2017.csv")
display(raw_df.head())
print('Number of observations: ',len(raw_df))

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue        G           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.54          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.80      3.79         3.29         35.57          82.0   
3     15.61     10.95      3.28         2.95         32.78          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count  User_Score  User_Count Rating  
0          51.0         8.0       324.0      E  
1           NaN         NaN         NaN    NaN  
2          73.0         8.3       712.0      E  
3          73.0         8.0       193.0      E  
4           NaN         NaN         NaN    NaN

Number of observations:  17416


## General Data Cleaning

### Checking for `NaN`s

In [2]:
raw_df.isnull().any()

Name               False
Platform           False
Year_of_Release     True
Genre              False
Publisher           True
NA_Sales           False
EU_Sales           False
JP_Sales           False
Other_Sales        False
Global_Sales       False
Critic_Score        True
Critic_Count        True
User_Score          True
User_Count          True
Rating              True
dtype: bool

//unsure what to do with ratings

The relevant variables with null values are `Year_of_Release` and `Publisher`. Let us first get the observations that have a null value for `Year_of_Release`.

In [3]:
null_year_df = raw_df[raw_df['Year_of_Release'].isnull()]
display(null_year_df)

Name Platform  Year_of_Release  \
12795                Mobile Ops: The One Year War     X360              NaN   
14210                                Housekeeping       DS              NaN   
14532            Fullmetal Alchemist: Brotherhood      PSP              NaN   
15162                 Wii de Asobu: Metroid Prime      Wii              NaN   
15784  Writing and Speaking Beautiful Japanese DS       DS              NaN   
16176                Ferrari: The Race Experience      Wii              NaN   
16953                                     Sumioni      PSV              NaN   
17271                                Free Running      PSP              NaN   

            Genre                 Publisher  NA_Sales  EU_Sales  JP_Sales  \
12795  Simulation        Namco Bandai Games       0.0      0.00      0.06   
14210      Action                   Unknown       0.0      0.00      0.04   
14532      Action                   Unknown       0.0      0.00      0.03   
15162     Shooter                  Nintendo       0.0      0.00      0.02   
15784        Misc                   Unknown       0.0      0.00      0.02   
16176      Racing  System 3 Arcade Software       0.0      0.01      0.00   
16953      Action                       NaN       0.0      0.00      0.01   
17271      Sports        Reef Entertainment       0.0      0.00      0.00   

       Other_Sales  Global_Sales  Critic_Score  Critic_Count  User_Score  \
12795          0.0          0.06           NaN           NaN         NaN   
14210          0.0          0.04           NaN           NaN         NaN   
14532          0.0          0.03           NaN           NaN         NaN   
15162          0.0          0.02           NaN           NaN         NaN   
15784          0.0          0.02           NaN           NaN         NaN   
16176          0.0          0.02          54.0           4.0         NaN   
16953          0.0          0.01           NaN           NaN         NaN   
17271          0.0          0.01           NaN           NaN         NaN   

       User_Count Rating  
12795         NaN    NaN  
14210         NaN    NaN  
14532         NaN    NaN  
15162         NaN    NaN  
15784         NaN    NaN  
16176         NaN      E  
16953         NaN    NaN  
17271         NaN      T

There are 8 observations with a null value for `Year_of_Release`. This seems to be a data collection error, since searching for these titles in the [VGChartz website](https://www.vgchartz.com/gamedb/) have release dates listed for each of them. Since we have access to the source of the dataset and there are only a few with this issue, we can simply add the correct value for each of them.

References: <br>
[Mobile Ops: The One Year War](https://www.vgchartz.com/game/24246/mobile-ops-the-one-year-war/?region=All) <br>
[Housekeeping](https://www.vgchartz.com/game/3497/housekeeping/?region=All) <br>
[Fullmetal Alchemist: Brotherhood](https://www.vgchartz.com/game/37307/fullmetal-alchemist-brotherhood/?region=All) <br>
[Wii de Asobu: Metroid Prime](https://www.vgchartz.com/game/27348/wii-de-asobu-metroid-prime/?region=All) <br>
[Writing and Speaking Beautiful Japanese DS](https://www.vgchartz.com/games/commentwall.php?id=4750) <br>
[Ferrari: The Race Experience](https://www.vgchartz.com/game/47029/ferrari-the-race-experience/?region=All) <br>
[Free Running](https://www.vgchartz.com/game/22809/free-running/?region=All)

In [4]:
raw_df.loc[12795,'Year_of_Release'] = 2008
raw_df.loc[14210,'Year_of_Release'] = 2006
raw_df.loc[14532,'Year_of_Release'] = 2009
raw_df.loc[15162,'Year_of_Release'] = 2009
raw_df.loc[15784,'Year_of_Release'] = 2008
raw_df.loc[16176,'Year_of_Release'] = 2010
raw_df.loc[17271,'Year_of_Release'] = 2007

Only one video game, [Sumioni](https://www.vgchartz.com/game/63722/sumioni/?region=All), does not have a listed release date. Coincidentally, it is also the game that has a null value in the `Publisher` variable. Thus, we can replace its value with the mean `Year_of_Release` and add the correct value for its `Publisher`.

In [5]:
raw_df.loc[16953,'Year_of_Release'] = round(raw_df['Year_of_Release'].mean(), 0)
raw_df.loc[16953,'Publisher'] = 'Acquire'

Since there are no longer missing values in `Year_of_Release`, we can convert the variable values into an integer.

In [6]:
raw_df['Year_of_Release'] = raw_df['Year_of_Release'].astype(np.int64)
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17416 entries, 0 to 17415
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             17416 non-null  object 
 1   Platform         17416 non-null  object 
 2   Year_of_Release  17416 non-null  int64  
 3   Genre            17416 non-null  object 
 4   Publisher        17416 non-null  object 
 5   NA_Sales         17416 non-null  float64
 6   EU_Sales         17416 non-null  float64
 7   JP_Sales         17416 non-null  float64
 8   Other_Sales      17416 non-null  float64
 9   Global_Sales     17416 non-null  float64
 10  Critic_Score     8336 non-null   float64
 11  Critic_Count     8336 non-null   float64
 12  User_Score       7798 non-null   float64
 13  User_Count       7798 non-null   float64
 14  Rating           10252 non-null  object 
dtypes: float64(9), int64(1), object(5)
memory usage: 2.0+ MB


### Checking for Inconsistent Categorical Values

The dataset has 3 categorical variables that have limited values: `Platform`, `Genre`, and `Rating`.

#### `Platform` variable

In [7]:
raw_df['Platform'].unique()

array(['Wii', 'NES', 'G', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       '3DS', 'PS4', 'N64', 'PS', 'X', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

The unique values of the `Platform` variable represent the platform codes in [VGChartz](https://www.vgchartz.com/charts/platform_totals/Hardware.php/). They are all valid. <br>
<i>Note: `G` and `X` represents `GB` and `XB`, respectively. </i>

#### `Genre` variable

In [18]:
raw_df['Genre'].unique()

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy'], dtype=object)

There are no misspelled values and are all valid genres listed in VGChartz.

#### `Rating` Variable

In [21]:
raw_df['Rating'].unique()

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

There are currently 7 ESRB ratings according to [IGN](https://www.ign.com/wikis/content-ratings/ESRB). All of the `Rating` non-null values are valid rating summaries apart from `K-A`. `K-A` (Kids-Adults) is the old term for the `E` (Everyone) rating. Thus, `K-A` values will be replaced by `E`.

In [23]:
raw_df['Rating'] = raw_df['Rating'].replace({'K-A': 'E'})
raw_df['Rating'].unique()

array(['E', nan, 'M', 'T', 'E10+', 'AO', 'EC', 'RP'], dtype=object)

### Checking for Duplicates

Let us first check if there are any duplicate `Name` values.

In [24]:
duplicate_names = raw_df['Name'].value_counts().reset_index(name='Count').query('Count > 1').sort_values(['Count', 'index'])
display(duplicate_names)
print('Duplicate Names: ', len(duplicate_names))
print('Total Number of Observations with Duplicate Names: ', duplicate_names['Count'].sum())

index  Count
2640                       007: The World is not Enough      2
2141                                  11eyes: CrossOver      2
1635                   18 Wheeler: American Pro Trucker      2
2652                                   187: Ride or Die      2
1994  2 in 1 Combo Pack: Sonic Heroes / Super Monkey...      2
...                                                 ...    ...
4                              LEGO Marvel Super Heroes      9
1                                         Madden NFL 07      9
2                                         Madden NFL 08      9
5                                           Ratatouille      9
0                           Need for Speed: Most Wanted     12

[2896 rows x 2 columns]

Duplicate Names:  2896
Total Number of Observations with Duplicate Names:  8232


There are 2,896 video game titles that have more than one observation and a total of 8,232 observations with duplicate titles. Let's take a closer look at the observations with the most duplicated title: `Need for Speed: Most Wanted`.

In [10]:
need_for_speed_observations = raw_df[raw_df['Name'] == 'Need for Speed: Most Wanted'].sort_values('Year_of_Release')
need_for_speed_observations

Name Platform  Year_of_Release   Genre  \
259    Need for Speed: Most Wanted      PS2             2005  Racing   
1601   Need for Speed: Most Wanted     X360             2005  Racing   
2007   Need for Speed: Most Wanted        X             2005  Racing   
3603   Need for Speed: Most Wanted       GC             2005  Racing   
6005   Need for Speed: Most Wanted       PC             2005  Racing   
6444   Need for Speed: Most Wanted       DS             2005  Racing   
6509   Need for Speed: Most Wanted      GBA             2005  Racing   
529    Need for Speed: Most Wanted      PS3             2012  Racing   
1198   Need for Speed: Most Wanted     X360             2012  Racing   
2036   Need for Speed: Most Wanted      PSV             2012  Racing   
11780  Need for Speed: Most Wanted       PC             2012  Racing   
6308   Need for Speed: Most Wanted     WiiU             2013  Racing   

             Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
259    Electronic Arts      2.03      1.79      0.08         0.47   
1601   Electronic Arts      1.00      0.13      0.02         0.10   
2007   Electronic Arts      0.53      0.46      0.00         0.05   
3603   Electronic Arts      0.43      0.11      0.00         0.02   
6005   Electronic Arts      0.02      0.23      0.00         0.04   
6444   Electronic Arts      0.24      0.01      0.00         0.02   
6509   Electronic Arts      0.19      0.07      0.00         0.00   
529    Electronic Arts      0.71      1.46      0.06         0.58   
1198   Electronic Arts      0.62      0.78      0.01         0.15   
2036   Electronic Arts      0.33      0.46      0.01         0.22   
11780  Electronic Arts      0.00      0.06      0.00         0.02   
6308   Electronic Arts      0.13      0.12      0.00         0.02   

       Global_Sales  Critic_Score  Critic_Count  User_Score  User_Count Rating  
259            4.37          82.0          36.0         9.1       138.0      T  
1601           1.25          83.0          54.0         8.4       135.0      T  
2007           1.04          83.0          32.0         8.8        29.0      T  
3603           0.56          80.0          18.0         9.1        23.0      T  
6005           0.29          82.0          19.0         8.5       531.0      T  
6444           0.27          45.0           4.0         6.1        22.0      E  
6509           0.26           NaN           NaN         8.3        14.0      E  
529            2.81           NaN           NaN         NaN         NaN    NaN  
1198           1.56          83.0          54.0         8.4       135.0      T  
2036           1.02           NaN           NaN         NaN         NaN    NaN  
11780          0.08          82.0          19.0         8.5       531.0      T  
6308           0.27           NaN           NaN         NaN         NaN    NaN

Closely examining the observations above reveals that the observations with the title `Need for Speed: Most Wanted` are similar in `Name`, `Genre`, and `Publisher`. Meanwhile, their values in the `Platform` and `Year_of_Release` columns vary. This may be due to the fact that the same game can be published across different platforms, which could affect the game's performance and sales. Some games may also receive remakes with the same name, which is considered to be a different game from the original. As such, observations with duplicate titles will be considered as separate observations.

With these in mind, let's check if there are any duplicate observations with the five aforementioned variables.

In [11]:
duplicate_rows = raw_df[['Name', 'Platform', 'Genre', 'Publisher', 'Year_of_Release']].value_counts().reset_index(name='Count').query('Count > 1').sort_values(['Count', 'Name'])
display(duplicate_rows)

print('Duplicate Observations: ', len(duplicate_rows))
print('Total Number of Observations with Duplicates: ', duplicate_rows['Count'].sum())

Name Platform     Genre        Publisher  Year_of_Release  \
1       Madden NFL 13      PS3    Sports  Electronic Arts             2012   
0  Sonic the Hedgehog      PS3  Platform             Sega             2006   

   Count  
1      2  
0      2

Duplicate Observations:  2
Total Number of Observations with Duplicates:  4


Now there are only two duplicates, `Madden NFL 13` and `	Sonic the Hedgehog`. Let's examine these observations.

In [12]:
duplicate_df = pd.merge(duplicate_rows.reset_index(), raw_df, how='inner').set_index('index')
duplicate_df.drop(['Count'], axis=1, inplace=True)
display(duplicate_df)

Name Platform     Genre        Publisher  \
index                                                           
1           Madden NFL 13      PS3    Sports  Electronic Arts   
1           Madden NFL 13      PS3    Sports  Electronic Arts   
0      Sonic the Hedgehog      PS3  Platform             Sega   
0      Sonic the Hedgehog      PS3  Platform             Sega   

       Year_of_Release  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
index                                                               
1                 2012      2.12      0.22      0.00         0.23   
1                 2012      0.00      0.01      0.00         0.00   
0                 2006      0.41      0.06      0.04         0.66   
0                 2006      0.00      0.48      0.00         0.00   

       Global_Sales  Critic_Score  Critic_Count  User_Score  User_Count Rating  
index                                                                           
1              2.57          83.0          22.0         5.5       101.0      E  
1              0.01          83.0          22.0         5.5       101.0      E  
0              1.16          43.0          17.0         4.1       177.0   E10+  
0              0.48          43.0          17.0         4.1       177.0   E10+

Both games have similar values to their duplicate apart from their total sales. As such, we will combine the sales of each game and replace the old values in the dataframe with the new, combined ones.

In [13]:
duplicate_title_observations = raw_df[raw_df['Name'].isin(duplicate_titles['index'].tail(20))].sort_values('Name')
duplicate_title_observations

Name Platform  Year_of_Release     Genre  \
8658      Angry Birds Star Wars     XOne             2013  Strategy   
10816     Angry Birds Star Wars     WiiU             2013  Strategy   
6077      Angry Birds Star Wars      PS3             2013  Strategy   
11392     Angry Birds Star Wars      PSV             2013  Strategy   
6472      Angry Birds Star Wars      Wii             2013  Strategy   
...                         ...      ...              ...       ...   
3148   The LEGO Movie Videogame     WiiU             2014    Action   
2194   The LEGO Movie Videogame      PS3             2014    Action   
5967   The LEGO Movie Videogame      PSV             2014    Action   
2954   The LEGO Movie Videogame      3DS             2014    Action   
12517  The LEGO Movie Videogame       PC             2014    Action   

                                    Publisher  NA_Sales  EU_Sales  JP_Sales  \
8658                               Activision      0.11      0.04      0.00   
10816                              Activision      0.05      0.04      0.00   
6077                               Activision      0.10      0.14      0.00   
11392                              Activision      0.03      0.03      0.00   
6472                               Activision      0.15      0.09      0.00   
...                                       ...       ...       ...       ...   
3148   Warner Bros. Interactive Entertainment      0.33      0.23      0.03   
2194   Warner Bros. Interactive Entertainment      0.33      0.45      0.02   
5967   Warner Bros. Interactive Entertainment      0.04      0.19      0.00   
2954   Warner Bros. Interactive Entertainment      0.28      0.31      0.05   
12517  Warner Bros. Interactive Entertainment      0.02      0.04      0.00   

       Other_Sales  Global_Sales  Critic_Score  Critic_Count  User_Score  \
8658          0.02          0.16          53.0           4.0         6.9   
10816         0.01          0.10           NaN           NaN         6.8   
6077          0.05          0.29          49.0           6.0         3.6   
11392         0.02          0.08           NaN           NaN         2.1   
6472          0.02          0.26           NaN           NaN         2.8   
...            ...           ...           ...           ...         ...   
3148          0.05          0.64          68.0           5.0         7.7   
2194          0.15          0.95          80.0           5.0         6.8   
5967          0.07          0.29           NaN           NaN         5.7   
2954          0.05          0.69           NaN           NaN         5.4   
12517         0.00          0.06          68.0           7.0         6.8   

       User_Count Rating  
8658         63.0      E  
10816        17.0      E  
6077         23.0      E  
11392        14.0      E  
6472          5.0      E  
...           ...    ...  
3148         25.0   E10+  
2194         12.0   E10+  
5967         15.0   E10+  
2954         13.0   E10+  
12517        33.0   E10+  

[169 rows x 15 columns]

In [14]:
duplicate_rows = raw_df[raw_df.duplicated()]
print('Number of duplicate observations: ', len(duplicate_rows))

Number of duplicate observations:  0


With this, we can conclude that every observation is a unique entry, thus there is no need for removing duplicates.